stripe-core/src/main/java/com/stripe/android/core/browser/customtabs/CustomTabsManager.kt

```diff
@@ -0,0 +1,221 @@
+package com.stripe.android.core.browser.customtabs
+
+import android.app.Activity
+import android.content.ActivityNotFoundException
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.os.Build
+import android.os.Build.VERSION_CODES
+import androidx.annotation.RestrictTo
+import androidx.browser.customtabs.CustomTabsCallback
+import androidx.browser.customtabs.CustomTabsClient
+import androidx.browser.customtabs.CustomTabsIntent
+import androidx.browser.customtabs.CustomTabsServiceConnection
+import androidx.browser.customtabs.CustomTabsSession
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import com.stripe.android.core.Logger
+import javax.inject.Inject
+
+/**
+ * Custom Tabs Service Connection manager implementation.
+ *
+ * Usage:
+ *
+ * 1. Ensure you register the CustomTabsService in your SDK's AndroidManifest.
+ * 2. Register the CustomTabsManager as a lifecycle observer in your activity.
+ * 3. (Optional) Call [CustomTabsManager.mayLaunchUrl] to pre-fetch the url.
+ * 4. Call [CustomTabsManager.openCustomTab] to open URLs.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+interface CustomTabsManager : DefaultLifecycleObserver {
+
+    /**
+     * Opens the URL on a Custom Tab if possible.
+     *
+     * @param activity The host activity.
+     * @param uri the Uri to be opened.
+     * @param fallback a CustomTabFallback to be used if Custom Tabs is not available.
+     */
+    fun openCustomTab(
+        activity: Activity,
+        uri: Uri,
+        fallback: () -> Unit
+    )
+
+    /**
+     * Warms up the browser for a given URL, so that it loads faster when launched.
+     */
+    fun mayLaunchUrl(url: String): Boolean
+}
+
+/**
+ * Manages the connection to the CustomTabsService.
+ *
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class CustomTabsManagerImpl @Inject constructor(
+    private val logger: Logger,
+    private val getCustomTabsPackage: GetCustomTabsPackage
+) : CustomTabsManager {
+
+    private var client: CustomTabsClient? = null
+    private var connection: CustomTabsServiceConnection? = null
+    private var session: CustomTabsSession? = null
+    private var serviceBound = false
+
+    /**
+     * Binds the Activity to the CustomTabsService.
+     */
+    override fun onStart(owner: LifecycleOwner) {
+        super.onStart(owner)
+        val context = owner as? Context ?: return
+        if (client != null) return
+        connection = ServiceConnection(object : ServiceConnectionCallback {
```

Comment:
Does this _need_ to happen in onStart/onStop or can this stay alive the entire lifecycle?

Reply:
followed the approach in the official Chromium CT example, where connection gets destroyed and recreated on start / stop: https://github.com/GoogleChrome/android-browser-helper/blob/7a2953b33daf921a6fe5e576d22e2af7664a2743/demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/CustomTabActivityHelper.java#L63-L73